test/quantized_ops/test_quantized_matmul.py

```diff
@@ -146,6 +177,54 @@ def test_int4_per_channel_linear_module(self):
       self.assertGreater(
           self._calc_cosine_dist(out_quant_xla.cpu(), out_quant), 0.999999)
 
+  def test_blockwise_matmul_op(self):
+    input_features = 6
+    out_features = 8
+    block_size = 2
+    batch_size = 3
+    for n_bit in [4]:
+      with self.subTest(n_bit=n_bit):
+        weight = torch.randint(-8, 7, (input_features // block_size, block_size,
+                                       out_features)).to(torch.int8)
+        weight_scaler = torch.ones(input_features // block_size, out_features)
+        x = torch.rand(batch_size, input_features)
+
+        # Fake quantize output.
+        w_dq = (weight * weight_scaler.unsqueeze(1)).reshape(
+            input_features, out_features)
+        fake_quant_out = torch.matmul(x, w_dq)
+        # Eager output.
+        torch_out = torch.ops.xla.quantized_matmul(
+            x, weight, weight_scaler, block_size=block_size)
+        self.assertGreater(
+            self._calc_cosine_dist(fake_quant_out, torch_out), 0.99999)
+        # XLA Output.
+        if not (n_bit == 4 and xr.device_type() != 'TPU'):
+          x = x.to(device)
+          weight = weight.to(device)
+          weight_scaler = weight_scaler.to(device)
+          xla_out = torch.ops.xla.quantized_matmul(
+              x, weight, weight_scaler, block_size=block_size)
+          self.assertTrue(torch.allclose(torch_out, xla_out.cpu(), atol=0.03))
+
+  def test_blockwise_linear_module(self):
+    for n_bit in [4, 8]:
+      with self.subTest(n_bit=n_bit):
+        m = M(6, 8)
+        x = torch.randn(3, 6)
+        out_fp = m(x)
+        m.replace_with_xla_quantized_matmul(n_bit=8, block_size=2)
+        out_quant = m(x)
+        self.assertGreater(self._calc_cosine_dist(out_fp, out_quant), 0.99)
+
+        # Dot with int4 weight is only supported on TPU
+        if not (n_bit == 4 and xr.device_type() != 'TPU'):
+          m = m.to(device)
```

Comment:
what's the behavior on `CUDA` device?

Reply:
Because int4 only runs on XLA:TPU

Reply:
offline discussion summary: `int4` works only on XLA:TPU today; XLA:CPU does not support quant ops at the moment. XLA:GPU level of support is unclear as it is not tested currently.